### PR TITLE
Add user defined labels to tracer

### PR DESCRIPTION
**Pull request by razlo**

### CODE REVIEW

**Comment by ranrib** [comment]
readme please @razlo

File: tracer/tracer.go [hunk 1]
```diff
@@ -255,6 +287,36 @@ func AddEvent(event *protocol.Event) {
...
+		valueSize = strconv.IntSize
+	case string:
+		valueSize = len(label.value.(string))
+	default:
+		if tracer.Config.Debug {
+			log.Println("EPSAGON DEBUG: type %v is not supported for labels. Supported types are: int, float, string", label.value)
+		}
+		return false
+	}
+	if len(label.key)+valueSize+tracer.labelsSize > MaxLabelSize {
```
Comment (maorlx) [inline tracer/tracer.go:303]: would rename MaxLabelSize to MaxLabelsSize (Label -> Labels)
add spaces between the +

File: tracer/tracer.go [hunk 2]
```diff
@@ -110,6 +126,20 @@ func HandleSendTracesResponse(resp *http.Response, err error) {
 
 func (tracer *epsagonTracer) getTraceReader() (io.Reader, error) {
 	version := "go " + runtime.Version()
+
+	for _, event := range tracer.events {
+		if event.Origin == "runner" {
+			jsonString, err := json.Marshal(tracer.labels)
```
Comment (maorlx) [inline tracer/tracer.go:132]: should break the loop after the runner event has found (alternatively - add some getRunnerEvent func)

File: tracer/tracer.go [hunk 3]
```diff
@@ -110,6 +126,20 @@ func HandleSendTracesResponse(resp *http.Response, err error) {
...
+			if err != nil {
+				if tracer.Config.Debug {
+					log.Printf("EPSAGON DEBUG failed appending labels")
```
Comment (maorlx) [inline tracer/tracer.go:135]: would move all the labels json logic to a separated func

File: tracer/tracer.go [hunk 4]
```diff
@@ -23,11 +26,16 @@ var (
 	GlobalTracer Tracer
 )
 
+// MaxLabelSize is the maximum allowed total labels size
+const MaxLabelSize = 10 * 1024
```
Comment (maorlx) [inline tracer/tracer.go:30]: how is this value determined? (is it like other insts?)

File: tracer/tracer.go [hunk 5]
```diff
@@ -255,6 +287,36 @@ func AddEvent(event *protocol.Event) {
 	GlobalTracer.AddEvent(event)
 }
 
+func (tracer *epsagonTracer) verifyLabel(label EpsagonLabel) bool {
+	var valueSize = 0
+	switch label.value.(type) {
+	case int, float64:
```
Comment (maorlx) [inline tracer/tracer.go:293]: maybe open a ticket to support more types? (uint, byte, maybe slices)

File: epsagon/common_utils.go [hunk 6]
```diff
@@ -1,7 +1,10 @@
 package epsagon
```
Comment (maorlx) [inline epsagon/common_utils.go:1]: dont forget to update:
- example (under examples dir)
- README
- docs

File: tracer/tracer.go (hunk 4 above)
Comment (razlo) [reply tracer/tracer.go:30]: yep

File: epsagon/common_utils.go [hunk 7]
```diff
@@ -30,3 +33,9 @@ func NewTracerConfig(applicationName, token string) *Config {
 		},
 	}
 }
+
+// Label adds a label to the sent trace
+func Label(key string, value interface{}, args ...context.Context) {
+	currentTracer := internal.ExtractTracer(args)
```
[the remark this answers is not in the document]
Comment (razlo) [reply epsagon/common_utils.go:39]: already checking. where do u mean?

[the remark this answers is not in the document]
Comment (maorlx) [reply epsagon/common_utils.go:39]: comment from an old version, irrelevant :)

File: tracer/tracer.go [hunk 8]
```diff
@@ -255,6 +291,45 @@ func AddEvent(event *protocol.Event) {
...
+		valueSize = strconv.IntSize
+	case string:
+		valueSize = len(label.value.(string))
+	default:
+		if tracer.Config.Debug {
+			log.Println("EPSAGON DEBUG: Supported label types are: int, float, string")
+		}
+		return false
+	}
+	if len(label.key)+valueSize+tracer.labelsSize > MaxLabelsSize {
```
Comment (maorlx) [inline tracer/tracer.go:307]: add spaces between +

File: README.md [hunk 9]
```diff
@@ -36,6 +38,26 @@ Or using `dep`:
...
+epsagon.Label('user_id', user_id)
+```
+
+You can also use it to ship custom metrics:
+```go
+epsagon.Label('key', 'metric')
+epsagon.Label('items_in_cart', items_in_cart)
+```
+
+Valid types are `string`, `bool`, `int` and `float`.
```
Comment (maorlx) [inline README.md:59]: we should add support to bool : ) (or remove it from here)

File: epsagon/common_utils.go [hunk 10]
```diff
@@ -1,7 +1,10 @@
 package epsagon
```
Comment (maorlx) [inline epsagon/common_utils.go:2]: add example (new example/add to existing example code)

**Review by maorlx — APPROVED** [review]

**Pull request merged**